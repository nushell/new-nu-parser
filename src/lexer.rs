use crate::compiler::Span;
use logos::Logos;

/// Average number of bytes per token used for estimating the tokens buffer size.
///
/// Estimated with this snippet:
/// let res = ls tests/**/*.nu | get name | each {|name|
///     let nbytes = open --raw $name | into binary | length
///     let ntokens = cargo run -- $name | lines | where $it starts-with 'Token3' | length
///     {
///         file: $name
///         nbytes: $nbytes
///         ntokens: $ntokens
///         bytes_per_token: ($nbytes / $ntokens)
///     }
/// }
///
/// TODO: Use larger and more representative codebase to estimate this
const AVG_BYTES_PER_TOKEN: usize = 2;

/// Lexed tokens
///
/// Tokens and spans are stored in separate vectors indexed by a position index (starting at 0).
#[derive(Clone)]
pub struct Tokens {
    pos: usize,
    tokens: Vec<Token>,
    spans: Vec<Span>,
}

impl Tokens {
    /// Create a new Tokens with allocated storage for the tokens and spans
    pub fn new(source: &[u8]) -> Self {
        let estimated_num_tokens = source.len() / AVG_BYTES_PER_TOKEN;
        Tokens {
            pos: 0,
            tokens: Vec::with_capacity(estimated_num_tokens),
            spans: Vec::with_capacity(estimated_num_tokens),
        }
    }

    // Position-related methods

    /// Advance position to point at the next token
    ///
    /// Note that this can potentially point beyond the tokens if called enough times. The parser
    /// should correctly check for EOF and terminate without advancing further.
    pub fn advance(&mut self) {
        debug_assert!(self.pos < self.tokens.len());
        self.pos += 1;
    }

    /// Return current position
    pub fn pos(&self) -> usize {
        self.pos
    }

    /// Set current position
    pub fn set_pos(&mut self, pos: usize) {
        self.pos = pos;
    }

    // Adding and fetching tokens

    /// Push a spanned token to the internal storage
    pub fn push(&mut self, token: Token, span: Span) {
        self.tokens.push(token);
        self.spans.push(span);
    }

    /// Check the token at the current position
    pub fn peek(&self) -> (Token, Span) {
        (self.peek_token(), self.peek_span())
    }

    /// Same as peek() but return only the token
    pub fn peek_token(&self) -> Token {
        self.tokens[self.pos]
    }

    /// Same as peek() but return only the span
    pub fn peek_span(&self) -> Span {
        self.spans[self.pos]
    }

    // Printing

    /// Format the tokens into a human-readable output for debugging
    pub fn display(&self, source: &[u8]) -> String {
        let mut result = String::new();

        result.push_str("==== TOKENS ====\n");

        for (i, (token, span)) in self.tokens.iter().zip(self.spans.iter()).enumerate() {
            result.push_str(&format!(
                "Token3 {i:4}: {:25} span: {:4} .. {:4} '{}'\n",
                format!("{:?}", token),
                span.start,
                span.end,
                String::from_utf8_lossy(
                    source
                        .get(span.start..span.end)
                        .expect("missing source of token span")
                )
                .replace("\r", "\\r")
                .replace("\n", "\\n")
                .replace("\t", "\\t")
            ));
        }

        result
    }

    /// Print the output of display() to standard output
    pub fn print(&self, source: &[u8]) {
        let output = self.display(source);
        print!("{output}");
    }

    /// Print the output of display() to standard error
    pub fn eprint(&self, source: &[u8]) {
        let output = self.display(source);
        eprint!("{output}");
    }
}

/// Lex the source contents and return allocated Tokens.
///
/// In the case of error, you can look up the last stored token to get a clue what went wrong. The
/// last stored token is always End Of File (EOF), so there will always be at least one token.
pub fn lex(contents: &[u8], span_offset: usize) -> (Tokens, Result<(), ()>) {
    // TODO: We might require the contents to always end with a newline, in which case return an error
    let mut tokens = Tokens::new(contents);
    let lexer = Token::lexer(contents).spanned();

    for (res, span) in lexer {
        match res {
            Ok(token) => tokens.push(
                token,
                Span::new(span.start + span_offset, span.end + span_offset),
            ),
            Err(_) => {
                tokens.push(
                    Token::Eof,
                    Span::new(span.end + span_offset, span.end + span_offset),
                );
                return (tokens, Err(()));
            }
        }
    }

    tokens.push(
        Token::Eof,
        Span::new(contents.len() + span_offset, contents.len() + span_offset),
    );

    (tokens, Ok(()))
}

#[derive(Logos, Debug, Clone, Copy, PartialEq)]
#[logos(skip r"[ \t]+")]
#[logos(source = [u8])]
pub enum Token {
    #[regex("(?:0[xob])?[0-9][0-9_]*", priority = 10)]
    Int,
    #[regex(r"(:?[0-9][0-9_]*)*\.([0-9][0-9_]*)*(?:[eE][+-]?[0-9_]+)?")]
    Float,
    #[regex("\n|\r\n|\x0C")]
    Newline,
    #[regex(r#""(?:[^"\\]|\\["\\bnfrt])*""#)]
    DoubleQuotedString,
    #[regex(r#"'[^']*'"#)]
    SingleQuotedString,
    #[regex(r#"`[^`]*`"#)]
    BacktickBareword,
    // #[regex(r#"[ \t]+"#)]
    // HorizontalWhitespace,
    #[regex(r#"[0-9]{4}-[0-9]{2}-[0-9]{2}(?:T[0-9]{2}:[0-9]{2}:[0-9]{2}(?:\.[0-9]+)?)?(?:Z|[\+-][0-9]{2}:[0-9]{2})?"#)]
    Datetime,
    #[regex(r#"#[^\n]*"#, priority = 20)]
    Comment,
    // lower priority to avoid clashing with Int
    #[regex(r#"(_|[^\s[:punct:]])(#|_|[^\s[:punct:]])*"#, priority = 2)]
    Bareword,
    #[token("...")]
    DotDotDot,
    #[token("..")]
    DotDot,
    #[token(".", priority = 10)] // higher priority to avoid "." being tokenized as Float
    Dot,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("[")]
    LSquare,
    #[token("]")]
    RSquare,
    #[token("{")]
    LCurly,
    #[token("}")]
    RCurly,
    #[token("<=")]
    LessThanEqual,
    #[token("<")]
    LessThan,
    #[token(">=")]
    GreaterThanEqual,
    #[token(">")]
    GreaterThan,
    #[token("++")]
    PlusPlus,
    #[token("+=")]
    PlusEquals,
    #[token("+")]
    Plus,
    #[token("->")]
    ThinArrow,
    #[token("=>")]
    ThickArrow,
    #[token("-=")]
    DashEquals,
    #[token("-")]
    Dash,
    #[token("**")]
    AsteriskAsterisk,
    #[token("*=")]
    AsteriskEquals,
    #[token("*")]
    Asterisk,
    #[token("//")]
    ForwardSlashForwardSlash,
    #[token("/=")]
    ForwardSlashEquals,
    #[token("/")]
    ForwardSlash,
    #[token("==")]
    EqualsEquals,
    #[token("=~")]
    EqualsTilde,
    #[token("=")]
    Equals,
    #[token("::")]
    ColonColon,
    #[token(":")]
    Colon,
    #[token("$")]
    Dollar,
    #[token(";")]
    Semicolon,
    #[token("!=")]
    ExclamationEquals,
    #[token("!~")]
    ExclamationTilde,
    #[token("!")]
    Exclamation,
    #[token("&&")]
    AmpersandAmpersand,
    #[token("&")]
    Ampersand,
    #[token(",")]
    Comma,
    #[token("?")]
    QuestionMark,
    #[token("^")]
    Caret,
    #[token("@")]
    At,
    #[token("||")]
    PipePipe,
    #[token("|")]
    Pipe,
    #[token("o>")]
    OutGreaterThan,
    #[token("o>>")]
    OutGreaterGreaterThan,
    #[token("e>")]
    ErrGreaterThan,
    #[token("e>>")]
    ErrGreaterGreaterThan,
    #[token("o+e>")]
    OutErrGreaterThan,
    #[token("o+e>>")]
    OutErrGreaterGreaterThan,
    #[token("e>|")]
    ErrGreaterThanPipe,
    #[token("o+e>|")]
    OutErrGreaterThanPipe,
    /// End of file, doesn't match any syntax, but source code always end with it
    Eof,
}

#[cfg(test)]
mod test {
    /// Lexer tests useful for smaller sources, errors and corner cases
    use crate::compiler::Span;
    use crate::lexer::{lex, Token};

    fn test_lex(src: &[u8], expected_tokens: &[(Token, Span)], expected_result: Result<(), ()>) {
        let (mut actual_tokens, actual_result) = lex(src, 0);

        assert_eq!(expected_result, actual_result, "Lexing result mismatch");

        for (i, expected) in expected_tokens.iter().enumerate() {
            let actual = actual_tokens.peek();
            assert_eq!(expected, &actual, "Mismatch in token {}", i);
            actual_tokens.advance();
        }
    }

    fn span(start: usize, end: usize) -> Span {
        Span { start, end }
    }

    #[test]
    fn lex_last_eof() {
        test_lex(b"", &[(Token::Eof, span(0, 0))], Ok(()));
    }

    #[test]
    fn lex_unmatched_string() {
        // TODO: Make unmatched delimiters nicer
        test_lex(b"'unmatched string", &[(Token::Eof, span(17, 17))], Err(()));
    }
}
