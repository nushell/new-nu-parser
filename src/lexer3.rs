use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"[ \t]+")]
#[logos(source = [u8])]
pub enum Token3 {
    #[regex("(?:0[xob])?[0-9][0-9_]*")]
    Int,
    #[regex("(?:0[xob])?[0-9][0-9_]*[a-zA-Z]+")]
    IntWithUnit,
    // #[regex(r"(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?")]
    // Float,
    #[regex("\n|\r\n|\x0C")]
    Newline,
    #[regex(r#""(?:[^"\\]|\\["\\bnfrt])*""#)]
    DoubleQuotedString,
    #[regex(r#"'[^']*'"#)]
    SingleQuotedString,
    #[regex(r#"`[^`]*`"#)]
    BacktickBareword,
    // #[regex(r#"[ \t]+"#)]
    // HorizontalWhitespace,
    #[token(".")]
    Dot,
    #[regex(r#"[0-9]{4}-[0-9]{2}-[0-9]{2}(?:T[0-9]{2}:[0-9]{2}:[0-9]{2}(?:\.[0-9]+)?)?(?:Z|[\+-][0-9]{2}:[0-9]{2})?"#)]
    Datetime,
    #[regex(r#"(?:_|[^\s[:punct:]])+"#, priority = 1)]
    Bareword,
}

#[cfg(test)]
mod test {
    use crate::lexer3::Token3;
    use logos::{Logos, Span};

    fn test_lex(src: &[u8], expected: &[Option<(Result<Token3, ()>, Span)>]) {
        let mut lexer = Token3::lexer(src).spanned();

        for exp in expected.iter() {
            let next = lexer.next();
            assert_eq!(exp, &next);

            if matches!(next, Some((Err(..), ..))) {
                return;
            }
        }

        assert_eq!(None, lexer.next());
    }

    fn span(start: usize, end: usize) -> Span {
        Span { start, end }
    }

    #[test]
    fn lex_int() {
        // TODO: test failing cases
    }

    #[test]
    fn lex_bareword() {
        // TODO: test failing cases
    }

    #[test]
    fn lex_datetime() {
        // TODO: test failing cases
    }
}
