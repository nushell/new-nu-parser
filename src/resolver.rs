use crate::protocol::{Command, Declaration};
use crate::{
    compiler::Compiler,
    errors::{Severity, SourceError},
    parser::{AstNode, BlockId, NodeId},
};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub struct ScopeId(pub usize);

#[derive(Debug, PartialEq)]
pub enum FrameType {
    /// Default scope frame marking the scope of a block/closure
    Scope,
    /// Immutable frame brought in by an overlay
    Overlay,
    /// Mutable frame inserted after activating an overlay to prevent mutating the overlay frame
    Light,
}

#[derive(Debug)]
pub struct Frame {
    pub frame_type: FrameType,
    pub variables: HashMap<Vec<u8>, NodeId>,
    pub decls: HashMap<Vec<u8>, NodeId>,
    /// Node that defined the scope frame (e.g., a block or overlay)
    pub node_id: NodeId,
}

impl Frame {
    pub fn new(scope_type: FrameType, node_id: NodeId) -> Self {
        Frame {
            frame_type: scope_type,
            variables: HashMap::new(),
            decls: HashMap::new(),
            node_id,
        }
    }
}

#[derive(Debug)]
pub struct Variable {
    pub is_mutable: bool,
}

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub struct VarId(pub usize);

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub struct DeclId(pub usize);

/// Fields extracted from Resolver
pub struct NameBindings {
    pub scope: Vec<Frame>,
    pub scope_stack: Vec<ScopeId>,
    pub variables: Vec<Variable>,
    pub var_resolution: HashMap<NodeId, VarId>,
    pub decls: Vec<Box<dyn Command>>,
    pub decl_resolution: HashMap<NodeId, DeclId>,
    pub errors: Vec<SourceError>,
}

impl NameBindings {
    pub fn new() -> Self {
        Self {
            scope: vec![],
            scope_stack: vec![],
            variables: vec![],
            var_resolution: HashMap::new(),
            decls: vec![],
            decl_resolution: HashMap::new(),
            errors: vec![],
        }
    }
}

impl Default for NameBindings {
    fn default() -> Self {
        Self::new()
    }
}

pub struct Resolver<'a> {
    // Immutable reference to a compiler after the first parsing pass
    compiler: &'a Compiler,

    /// All scope frames ever entered, indexed by ScopeId
    pub scope: Vec<Frame>,
    /// Stack of currently entered scope frames
    pub scope_stack: Vec<ScopeId>,
    /// Variables, indexed by VarId
    pub variables: Vec<Variable>,
    /// Mapping of variable's name node -> Variable
    pub var_resolution: HashMap<NodeId, VarId>,
    /// Declarations (commands, aliases, etc.), indexed by DeclId
    pub decls: Vec<Box<dyn Command>>,
    /// Mapping of decl's name node -> Command
    pub decl_resolution: HashMap<NodeId, DeclId>,
    /// Errors encountered during name binding
    pub errors: Vec<SourceError>,
}

impl<'a> Resolver<'a> {
    pub fn new(compiler: &'a Compiler) -> Self {
        Self {
            compiler,
            scope: vec![],
            scope_stack: vec![],
            variables: vec![],
            var_resolution: HashMap::new(),
            decls: vec![],
            decl_resolution: HashMap::new(),
            errors: vec![],
        }
    }

    pub fn to_name_bindings(self) -> NameBindings {
        NameBindings {
            scope: self.scope,
            scope_stack: self.scope_stack,
            variables: self.variables,
            var_resolution: self.var_resolution,
            decls: self.decls,
            decl_resolution: self.decl_resolution,
            errors: self.errors,
        }
    }

    pub fn print(&self) {
        let output = self.display_state();
        print!("{output}");
    }

    #[allow(clippy::format_collect)]
    pub fn display_state(&self) -> String {
        let mut result = String::new();

        result.push_str("==== SCOPE ====\n");
        for (i, scope) in self.scope.iter().enumerate() {
            result.push_str(&format!(
                "{i}: Frame {0:?}, node_id: {1:?}",
                scope.frame_type, scope.node_id
            ));

            let mut vars: Vec<String> = scope
                .variables
                .iter()
                .map(|(name, id)| format!("{0}: {id:?}", String::from_utf8_lossy(name)))
                .collect();

            let mut decls: Vec<String> = scope
                .decls
                .iter()
                .map(|(name, id)| format!("{0}: {id:?}", String::from_utf8_lossy(name)))
                .collect();

            if vars.is_empty() && decls.is_empty() {
                result.push_str(" (empty)\n");
                continue;
            }

            result.push('\n');

            if !vars.is_empty() {
                vars.sort();
                let line_var = format!("  variables: [ {0} ]\n", vars.join(", "));
                result.push_str(&line_var);
            }

            if !decls.is_empty() {
                decls.sort();
                let line_decl = format!("      decls: [ {0} ]\n", decls.join(", "));
                result.push_str(&line_decl);
            }
        }

        if !self.errors.is_empty() {
            result.push_str("==== SCOPE ERRORS ====\n");
            for error in &self.errors {
                result.push_str(&format!(
                    "{:?} (NodeId {}): {}\n",
                    error.severity, error.node_id.0, error.message
                ));
            }
        }

        result
    }

    pub fn resolve(&mut self) {
        if !self.compiler.ast_nodes.is_empty() {
            let last = self.compiler.ast_nodes.len() - 1;
            let last_node_id = NodeId(last);
            self.resolve_node(last_node_id)
        }
    }

    pub fn resolve_node(&mut self, node_id: NodeId) {
        // TODO: Move node_id param to the end, same as in typechecker
        match self.compiler.ast_nodes[node_id.0] {
            AstNode::Variable => self.resolve_variable(node_id),
            AstNode::Call { ref parts } => self.resolve_call(node_id, parts),
            AstNode::Block(block_id) => self.resolve_block(node_id, block_id, None),
            AstNode::Closure { params, block } => {
                // making sure the closure parameters and body end up in the same scope frame
                let closure_scope = if let Some(params) = params {
                    self.enter_scope(block);
                    self.resolve_node(params);
                    Some(self.exit_scope())
                } else {
                    None
                };

                let AstNode::Block(block_id) = self.compiler.ast_nodes[block.0] else {
                    panic!("internal error: closure's body is not a block");
                };

                self.resolve_block(block, block_id, closure_scope);
            }
            AstNode::Def {
                name,
                params,
                return_ty: _,
                block,
            } => {
                // define the command before the block to enable recursive calls
                self.define_decl(name);

                // making sure the def parameters and body end up in the same scope frame
                self.enter_scope(block);
                self.resolve_node(params);
                let def_scope = self.exit_scope();

                let AstNode::Block(block_id) = self.compiler.ast_nodes[block.0] else {
                    panic!("internal error: command definition's body is not a block");
                };

                self.resolve_block(block, block_id, Some(def_scope));
            }
            AstNode::Params(ref params) => {
                for param in params {
                    if let AstNode::Param { name, .. } = self.compiler.ast_nodes[param.0] {
                        self.define_variable(name, false);
                    } else {
                        panic!("param is not a param");
                    }
                }
            }
            AstNode::Let {
                variable_name,
                ty: _,
                initializer,
                is_mutable,
            } => {
                self.resolve_node(initializer);
                self.define_variable(variable_name, is_mutable)
            }
            AstNode::While { condition, block } => {
                self.resolve_node(condition);
                self.resolve_node(block);
            }
            AstNode::For {
                variable,
                range,
                block,
            } => {
                // making sure the for loop variable and body end up in the same scope frame
                self.enter_scope(block);
                self.define_variable(variable, false);
                let for_body_scope = self.exit_scope();

                self.resolve_node(range);

                let AstNode::Block(block_id) = self.compiler.ast_nodes[block.0] else {
                    panic!("internal error: for's body is not a block");
                };

                self.resolve_block(block, block_id, Some(for_body_scope));
            }
            AstNode::Loop { block } => {
                self.resolve_node(block);
            }
            AstNode::BinaryOp { lhs, op: _, rhs } => {
                self.resolve_node(lhs);
                self.resolve_node(rhs);
            }
            AstNode::Range { lhs, rhs } => {
                self.resolve_node(lhs);
                self.resolve_node(rhs);
            }
            AstNode::List(ref nodes) => {
                for node in nodes {
                    self.resolve_node(*node);
                }
            }
            AstNode::Table { header, ref rows } => {
                self.resolve_node(header);
                for row in rows {
                    self.resolve_node(*row);
                }
            }
            AstNode::Record { ref pairs } => {
                for (key, val) in pairs {
                    self.resolve_node(*key);
                    self.resolve_node(*val);
                }
            }
            AstNode::MemberAccess { target, field } => {
                self.resolve_node(target);
                self.resolve_node(field);
            }
            AstNode::If {
                condition,
                then_block,
                else_block,
            } => {
                self.resolve_node(condition);
                self.resolve_node(then_block);
                if let Some(block) = else_block {
                    self.resolve_node(block);
                }
            }
            AstNode::Match {
                target,
                ref match_arms,
            } => {
                self.resolve_node(target);
                for (arm_lhs, arm_rhs) in match_arms {
                    self.resolve_node(*arm_lhs);
                    self.resolve_node(*arm_rhs);
                }
            }
            AstNode::Statement(node) => self.resolve_node(node),
            AstNode::Param { .. } => (/* seems unused for now */),
            AstNode::Type { .. } => ( /* probably doesn't make sense to resolve? */ ),
            AstNode::NamedValue { .. } => (/* seems unused for now */),
            // All remaining matches do not contain NodeId => there is nothing to resolve
            _ => (),
        }
    }

    pub fn resolve_variable(&mut self, unbound_node_id: NodeId) {
        let var_name = trim_var_name(self.compiler.get_span_contents(unbound_node_id));

        if let Some(node_id) = self.find_variable(var_name) {
            let var_id = self
                .var_resolution
                .get(&node_id)
                .expect("internal error: missing resolved variable");

            self.var_resolution.insert(unbound_node_id, *var_id);
        } else {
            self.errors.push(SourceError {
                message: format!("variable `{}` not found", String::from_utf8_lossy(var_name)),
                node_id: unbound_node_id,
                severity: Severity::Error,
            })
        }
    }

    pub fn resolve_call(&mut self, unbound_node_id: NodeId, parts: &[NodeId]) {
        // Find out the potentially longest command name
        let max_name_parts = parts
            .iter()
            .position(|part| matches!(self.compiler.ast_nodes[part.0], AstNode::Name))
            .expect("call does not have any name")
            + 1;

        // Try to find the longest matching subcommand
        let first_start = self.compiler.spans[parts[0].0].start;

        for n in (0..max_name_parts).rev() {
            let last_end = self.compiler.spans[parts[n].0].end;
            let name = self
                .compiler
                .get_span_contents_manual(first_start, last_end);

            if let Some(node_id) = self.find_decl(name) {
                let decl_id = self
                    .decl_resolution
                    .get(&node_id)
                    .expect("internal error: missing resolved decl");

                self.decl_resolution.insert(unbound_node_id, *decl_id);
                break;
            }
        }

        // TODO? If the call does not correspond to any existing decl, it is an external call

        // Resolve args
        for part in &parts[max_name_parts..] {
            self.resolve_node(*part);
        }
    }

    pub fn resolve_block(
        &mut self,
        node_id: NodeId,
        block_id: BlockId,
        reused_scope: Option<ScopeId>,
    ) {
        let block = self
            .compiler
            .blocks
            .get(block_id.0)
            .expect("internal error: missing block");

        if let Some(scope_id) = reused_scope {
            self.enter_existing_scope(scope_id);
        } else {
            self.enter_scope(node_id);
        }

        for inner_node_id in &block.nodes {
            self.resolve_node(*inner_node_id);
        }
        self.exit_scope();
    }

    /// Enter an existing scope frame, e.g., a block or a closure
    pub fn enter_existing_scope(&mut self, scope_id: ScopeId) {
        self.scope_stack.push(scope_id);
    }

    /// Enter a new scope frame, e.g., a block or a closure
    pub fn enter_scope(&mut self, node_id: NodeId) {
        self.scope.push(Frame::new(FrameType::Scope, node_id));
        self.scope_stack.push(ScopeId(self.scope.len() - 1));
    }

    /// Exit a scope frame, e.g., when reaching the end of a block or a closure
    ///
    /// When exiting a scope frame, all overlays (and corresponding light frames) are removed along
    /// with the removed frame.
    pub fn exit_scope(&mut self) -> ScopeId {
        match self
            .scope_stack
            .iter()
            .rposition(|scope_id| self.scope[scope_id.0].frame_type == FrameType::Scope)
        {
            None => panic!("internal error: no scope frame to exit"),
            Some(pos) => {
                let scope_id = self.scope_stack[pos];
                self.scope_stack.truncate(pos);
                scope_id
            }
        }
    }

    pub fn define_variable(&mut self, var_name_id: NodeId, is_mutable: bool) {
        let var_name = self.compiler.get_span_contents(var_name_id);
        let var_name = trim_var_name(var_name).to_vec();

        let current_scope_id = self
            .scope_stack
            .last()
            .expect("internal error: missing scope frame id");

        self.scope[current_scope_id.0]
            .variables
            .insert(var_name, var_name_id);

        let var = Variable { is_mutable };
        self.variables.push(var);
        let var_id = VarId(self.variables.len() - 1);

        // let the definition of a variable also count as its use
        self.var_resolution.insert(var_name_id, var_id);
    }

    pub fn define_decl(&mut self, decl_name_id: NodeId) {
        // TODO: Deduplicate code with define_variable()
        let decl_name = self.compiler.get_span_contents(decl_name_id);
        let decl_name = trim_decl_name(decl_name).to_vec();
        let decl = Declaration::new(String::from_utf8_lossy(&decl_name).to_string());

        let current_scope_id = self
            .scope_stack
            .last()
            .expect("internal error: missing scope frame id");

        self.scope[current_scope_id.0]
            .decls
            .insert(decl_name, decl_name_id);

        self.decls.push(Box::new(decl));
        let decl_id = DeclId(self.decls.len() - 1);

        // let the definition of a decl also count as its use
        self.decl_resolution.insert(decl_name_id, decl_id);
    }

    pub fn find_variable(&self, var_name: &[u8]) -> Option<NodeId> {
        for scope_id in self.scope_stack.iter().rev() {
            if let Some(id) = self.scope[scope_id.0].variables.get(var_name) {
                return Some(*id);
            }
        }

        None
    }

    pub fn find_decl(&self, var_name: &[u8]) -> Option<NodeId> {
        // TODO: Deduplicate code with find_variable()
        for scope_id in self.scope_stack.iter().rev() {
            if let Some(id) = self.scope[scope_id.0].decls.get(var_name) {
                return Some(*id);
            }
        }

        None
    }
}

fn trim_var_name(name: &[u8]) -> &[u8] {
    if name.starts_with(b"$") && name.len() > 1 {
        &name[1..]
    } else {
        name
    }
}
fn trim_decl_name(name: &[u8]) -> &[u8] {
    if (name.starts_with(b"'") && name.ends_with(b"'"))
        || (name.starts_with(b"\"") && name.ends_with(b"\""))
        || (name.starts_with(b"`") && name.ends_with(b"`"))
    {
        &name[1..name.len() - 1]
    } else {
        name
    }
}
