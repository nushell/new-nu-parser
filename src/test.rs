use crate::lexer::Lexer;
use crate::lexer3::Token3;
use crate::resolver::Resolver;
use crate::typechecker::Typechecker;
use crate::{compiler::Compiler, parser::Parser};

use logos::Logos;

use std::path::Path;

fn evaluate_example(fname: &Path) -> String {
    let mut compiler = Compiler::new();
    let contents = std::fs::read(fname).expect("We only run tests found by glob");

    let span_offset = compiler.span_offset();
    compiler.add_file(&fname.to_string_lossy(), &contents);

    let parser = Parser::new(compiler, span_offset);
    compiler = parser.parse();

    let mut result = compiler.display_state();

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut resolver = Resolver::new(&compiler);
    resolver.resolve();
    result.push_str(&resolver.display_state());

    compiler.merge_name_bindings(resolver.to_name_bindings());

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut typechecker = Typechecker::new(&compiler);
    typechecker.typecheck();
    result.push_str(&typechecker.display_state());

    compiler.merge_types(typechecker.to_types());

    result
}

fn evaluate_lexer(fname: &Path) -> String {
    let contents = std::fs::read(fname);

    let Ok(contents) = contents else {
        panic!("Lexer: can't find file {}", fname.to_string_lossy());
    };

    let mut lexer = Lexer::new(&contents, 0);
    lexer.lex();
    lexer.display_state()
}

fn evaluate_lexer3(fname: &Path) -> String {
    let contents = std::fs::read(fname);

    let Ok(contents) = contents else {
        panic!("Lexer: can't find file {}", fname.to_string_lossy());
    };

    let mut lexer3 = Token3::lexer(&contents).spanned();
    let mut tokens = Vec::with_capacity(contents.len());
    let mut i = 0;
    let mut res = String::new();
    while let Some((token, span)) = lexer3.next() {
        let Ok(tok) = token else {
            res.push_str(&format!(
                "Token {i:4}: {:25?} {:4} .. {:4}\n",
                token, span.start, span.end,
            ));
            break;
        };

        res.push_str(&format!(
            "Token {i:4}: {:25} span: {:4} .. {:4} '{}'\n",
            format!("{:?}", tok),
            span.start,
            span.end,
            String::from_utf8_lossy(
                contents
                    .get(span.start..span.end)
                    .expect("lexer: missing source for span"),
            )
            .replace("\r", "\\r")
            .replace("\n", "\\n")
            .replace("\t", "\\t")
        ));
        i += 1;
        tokens.push(tok);
    }

    res
}

#[test]
fn test_node_output() {
    insta::glob!("../tests", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_example(path));
    });
}

#[test]
fn test_lexer() {
    insta::glob!("../tests/lex", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_lexer(path));
    });
}

#[test]
fn test_lexer3() {
    insta::glob!("../tests/lex", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_lexer3(path));
    });
}
