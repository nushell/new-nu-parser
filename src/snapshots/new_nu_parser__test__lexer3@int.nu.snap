---
source: src/test.rs
expression: evaluate_lexer3(path)
input_file: tests/lex/int.nu
snapshot_kind: text
---
Token    0: Int                       span:    0 ..    1 '0'
Token    1: Newline                   span:    1 ..    2 '\n'
Token    2: Int                       span:    2 ..    4 '00'
Token    3: Newline                   span:    4 ..    5 '\n'
Token    4: Int                       span:    5 ..   10 '0x123'
Token    5: Newline                   span:   10 ..   11 '\n'
Token    6: Int                       span:   11 ..   16 '0b101'
Token    7: Newline                   span:   16 ..   17 '\n'
Token    8: Int                       span:   17 ..   19 '0_'
Token    9: Newline                   span:   19 ..   20 '\n'
Token   10: IntWithUnit               span:   20 ..   23 '0bo'
Token   11: Newline                   span:   23 ..   24 '\n'
Token   12: IntWithUnit               span:   24 ..   26 '0x'
Token   13: Newline                   span:   26 ..   27 '\n'
Token   14: Bareword                  span:   27 ..   32 '01x10'
Token   15: Newline                   span:   32 ..   33 '\n'
